### ISSUE TITLE
Trying to read `Context.Parser.Record` in `BadDataFound` callback causes stack overflow

### ISSUE THREAD

**Issue by tznind** [JoshClose/CsvHelper#1717]
**Describe the bug**
Accessing the `Record` property of the parser context in `BadDataFound` callback immediately triggers the callback to call again which results in a stack overflow.

**To Reproduce**

```csharp
public class TestStackOverflow
    {
        
        [Test]
        public void StackOverflow()
        {
            var badstring = new StringReader("Fish,\"DDDD");

            var cfg = new CsvConfiguration(CultureInfo.CurrentCulture)
            {
                BadDataFound = (s)=>{ Console.WriteLine($"Bad Data: {s.Context.Parser.Record}");}
            };
            var r = new CsvReader(badstring,cfg);
            Assert.DoesNotThrow(()=>r.Read());
        }
    }
```

Run test `dotnet test --filter TestStackOverflow`

```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
The active test run was aborted. Reason: Test host process crashed : Stack overf
low.


Test Run Aborted.
```

Affected version: `24.0.1`

**Comment by JoshClose** [comment]
Accessing `Parser[int]` or `Parser.Record` will cause the fields to be processed. When `BadDataFound` is invoked and the callback accesses those, it processes them again, and you get your overflow.

Do you any thoughts on a good solution for this? I can think of a couple things.

1. I could detect that it's being called and throw an exception letting the user know that you can't access those inside the callback.
1. I could save the fields that have already been processed so when detected it only returns processed fields up to the one that had the exception.

Something else?

**Comment by tznind** [comment]
Thanks for looking into this.  For me I think option 2 would make most sense although I'd be happy with it being null or empty too.

I encountered the overflow when updating from `19.0.0` so don't think previous versions had this issue so again 2 would be most consistent with the previous versions behavior.

**Comment by JoshClose** [comment]
Starting in 2.0.0, field detection happens for a row first. Only when a field is accessed is it processed and the bad data is detected. If you have 50 fields and only access 1, it shouldn't need to process the others.

I have the exception working, but I'll see about returning an empty string instead.

**Comment by JoshClose** [comment]
I'm not seeing a good way to be able to access the record. If you do `Parser[3]` and never accessed 0-2, there could be bad data in 0-2 also.

The args passed to `BadDataFound` contains the field that caused the problem. You can get the full row by doing `Context.Parser.RawRecord` which gives you the full string unchanged.

I think I'll have it throw an exception with a message stating you shouldn't use it and to use the field instead. Maybe I'll add `string RawRecord` to `BadDataFoundArgs` so people don't have to hunt for it. Maybe the index also.

**Comment by tznind** [comment]
Thanks for investigating, an Exception definetly beats a stack overflow!

**Comment by JoshClose** [comment]
Fix is in 25.0.0 on NuGet. https://www.nuget.org/packages/CsvHelper/25.0.0
